src/skeymgmt.c

```diff
@@ -0,0 +1,290 @@
+            slen -= tlen + 1;
+            str = p + 1;
+        } else {
+            tlen = slen;
+            slen = 0;
+            str = NULL;
+        }
+        if ((tlen == 7) && (strncmp(tok, "encrypt", tlen) == 0)) {
+            *flags |= CKF_ENCRYPT | CKF_DECRYPT;
+        } else if ((tlen == 3) && (strncmp(tok, "mac", tlen) == 0)) {
+            *flags |= CKF_SIGN | CKF_VERIFY;
+        } else if ((tlen == 4) && (strncmp(tok, "wrap", tlen) == 0)) {
+            *flags |= CKF_WRAP | CKF_UNWRAP;
+        } else if ((tlen == 6) && (strncmp(tok, "derive", tlen) == 0)) {
+            *flags |= CKF_DERIVE;
+        } else {
+            return CKR_ARGUMENTS_BAD;
+        }
+    }
+
+    return CKR_OK;
+}
+
+static void *p11prov_aes_generate(void *provctx, const OSSL_PARAM params[])
+{
+    P11PROV_CTX *ctx = (P11PROV_CTX *)provctx;
+    CK_MECHANISM mech = {
+        .mechanism = CKM_AES_KEY_GEN,
+        .pParameter = NULL,
+        .ulParameterLen = 0,
+    };
+    CK_ULONG keylen = 32; /* default to 256 bit */
+    CK_ATTRIBUTE tmpl[12] = {
+        { CKA_VALUE_LEN, &keylen, sizeof(keylen) },
+        { 0 },
+    };
+    size_t tmax = sizeof(tmpl) / sizeof(CK_ATTRIBUTE);
+    size_t tsize = 1;
+    P11PROV_URI *uri = NULL;
+    int ephemeral = 1; /* by default we create session keys */
+    /* all flags by default */
+    CK_FLAGS usage = CKF_ENCRYPT | CKF_DECRYPT | CKF_SIGN | CKF_VERIFY
+                     | CKF_WRAP | CKF_UNWRAP | CKF_DECRYPT;
+    const OSSL_PARAM *p;
+    CK_SLOT_ID slotid = CK_UNAVAILABLE_INFORMATION;
+    CK_RV rv = CKR_MECHANISM_PARAM_INVALID;
+    P11PROV_SESSION *session = NULL;
+    CK_OBJECT_HANDLE key_handle = CK_INVALID_HANDLE;
+    P11PROV_OBJ *key = NULL;
+    int ret;
+
+    P11PROV_debug("aes generate");
+
+    if (!ctx) {
+        return NULL;
+    }
+
+    /* CKA_ID and CKA_LABEL from template URI */
+    p = OSSL_PARAM_locate_const(params, P11PROV_PARAM_URI);
+    if (p) {
+        const char *puri = NULL;
+        ret = OSSL_PARAM_get_utf8_string_ptr(p, &puri);
+        if (ret != RET_OSSL_OK) {
+            P11PROV_raise(ctx, rv, "Invalid uri parameter type");
+            goto done;
+        }
+        uri = p11prov_parse_uri(ctx, puri);
+        if (!uri) {
+            goto done;
+        }
+
+        /* Id */
+        tmpl[tsize] = p11prov_uri_get_id(uri);
+        if (tmpl[tsize].ulValueLen != 0) tsize++;
+        /* Label */
+        tmpl[tsize] = p11prov_uri_get_label(uri);
+        if (tmpl[tsize].ulValueLen != 0) tsize++;
+    }
+
+    /* Key length */
+    p = OSSL_PARAM_locate_const(params, OSSL_SKEY_PARAM_KEY_LENGTH);
+    if (p) {
+        size_t len;
+        ret = OSSL_PARAM_get_size_t(p, &len);
+        if (ret != RET_OSSL_OK) {
+            P11PROV_raise(ctx, rv, "Invalid length parameter type");
+            goto done;
+        }
+        if (len != 16 && len != 24 && len != 32) {
+            P11PROV_raise(ctx, rv, "Invalid length parameter value");
```

Comment:
I would be more specific in the error messages. Using just `length` is quite generic here:
```suggestion
            P11PROV_raise(ctx, rv, "Invalid key length parameter type");
            goto done;
        }
        if (len != 16 && len != 24 && len != 32) {
            P11PROV_raise(ctx, rv, "Invalid key length parameter value");
```